### PR TITLE
WASM ABI: implement `row_iter_bsatn_advance` & `row_iter_bsatn_close`

### PR DESCRIPTION
**Pull request by Centril**
# Description of Changes

Implements `row_iter_bsatn_advance` & `row_iter_bsatn_close`.

cc https://github.com/clockworklabs/SpacetimeDB/issues/1460

### CODE REVIEW

File: crates/core/src/host/wasmtime/wasm_instance_env.rs [hunk 1]
```diff
@@ -582,69 +582,121 @@ impl WasmInstanceEnv {
...
+            let buffer_len = u32::read_from(mem, buffer_len_ptr)?;
+            // Get a mutable view to the `buffer`.
+            let mut buffer = mem.deref_slice_mut(buffer_ptr, buffer_len)?;
 
             let mut written = 0;
-            // Advance the iterator.
+            // Fill the buffer as much as possible.
             while let Some(chunk) = iter.as_slice().first() {
-                let Some(buf_chunk) = wasm_buffer.get_mut(..chunk.len()) else {
+                // TODO(Centril): refactor using `split_at_mut_checked`.
```
Comment (RReverser) [inline crates/core/src/host/wasmtime/wasm_instance_env.rs:639]: Alternatively, could leverage the `std::io::Write` implementation on slices. `.write()` on a slice results in the same zero-cost memcpy code and automatically takes care of "advancing" the slice to the remainder part: https://rust.godbolt.org/z/1x4W6KeYx

File: crates/bindings-csharp/Runtime/Internal/ITable.cs [hunk 2]
```diff
@@ -22,31 +22,45 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
...
+                        // Iterator exhausted, we're done.
+                        case -1:
+                            handle = FFI.RowIter.INVALID;
+                            return buffer_len != 0;
+                        case 0:
+                            return buffer_len != 0;
+                        case (short)(ushort)FFI.Errno.NO_SUCH_ITER:
+                            throw new NoSuchIterException();
+                        case (short)(ushort)FFI.Errno.BUFFER_TOO_SMALL:
+                            buffer = new byte[buffer_len];
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/ITable.cs:51]: Shouldn't this copy the original buffer? Not sure if any data is written to the buffer at that point or not.

File: crates/bindings-csharp/Runtime/Internal/FFI.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ public readonly record struct BytesSource(uint Handle)
 public readonly record struct BytesSink(uint Handle) { }
 
 #pragma warning disable IDE1006 // Naming Styles - Not applicable to FFI stuff.
-internal static partial class FFI
+public static partial class FFI
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/FFI.cs:22]: Is this just for Errno? Can we extract it like BytesSource above if so? I'd prefer not to expose APIs we can easily keep hidden.

File: crates/bindings-csharp/Runtime/Internal/FFI.cs [hunk 4]
```diff
@@ -159,14 +161,14 @@ out RowIter out_
     );
 
     [LibraryImport(StdbNamespace)]
-    public static partial CheckedStatus _iter_advance(
+    public static partial short _row_iter_bsatn_advance(
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/FFI.cs:164]: This can be just `Errno` now, right?

File: crates/bindings-csharp/Runtime/Internal/ITable.cs [hunk 5]
```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
...
+                {
+                    return false;
+                }
+
                 uint buffer_len;
                 while (true)
                 {
                     buffer_len = (uint)buffer.Length;
-                    try
+                    var ret = (FFI.Errno)
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/ITable.cs:34]: Then you don't need the cast here.

File: crates/bindings-csharp/Runtime/Internal/ITable.cs [hunk 6]
```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
...
+                        handle = FFI.RowIter.INVALID;
                     }
-                    catch (BufferTooSmallException)
+                    switch (ret)
                     {
-                        buffer = new byte[buffer_len];
-                        continue;
+                        // Iterator exhausted, we're done.
+                        case FFI.Errno.EXHAUSTED:
+                        // Iterator advanced but not exhausted,
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/ITable.cs:44]: I'm confused by this comment, is something missing? It's under `EXHAUSTED:` but says "iterator advanced but not exhausted". Also there is no `break;` for this `case` 🤔

File: crates/bindings-csharp/Runtime/Internal/ITable.cs [hunk 7]
```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
...
+                            Current = new byte[buffer_len];
+                            Array.Copy(buffer, 0, Current, 0, buffer_len);
+                            return buffer_len != 0;
+                        // Couldn't find the iterator, error!
+                        case FFI.Errno.NO_SUCH_ITER:
+                            throw new NoSuchIterException();
+                        // The scratch `buffer` is too small to fit a row / chunk.
+                        // Grow `buffer` and try again.
+                        case FFI.Errno.BUFFER_TOO_SMALL:
+                            buffer = new byte[buffer_len];
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/ITable.cs:57]: Unlike the comment says, it doesn't grow the `buffer` here?

Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/ITable.cs:57]: Ahh I see, `buffer_len` has a new length at this point. Might be worth adding to the comment, it's easy to miss.

File: crates/bindings-csharp/Runtime/Internal/ITable.cs [hunk 8]
```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
...
-                        continue;
+                        // Iterator exhausted, we're done.
+                        case FFI.Errno.EXHAUSTED:
+                        // Iterator advanced but not exhausted,
+                        // we'll need to advance the iterator in the next call to `MoveNext`.
+                        case FFI.Errno.OK:
+                            // Success! Copy over the row data to `Current` from the scratch `buffer`.
+                            Current = new byte[buffer_len];
+                            Array.Copy(buffer, 0, Current, 0, buffer_len);
+                            return buffer_len != 0;
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/ITable.cs:50]: FWIW `MoveNext()` should return `false` when the iterator is exhausted. Empty returned buffer used to indicate that before, but IIUC not anymore - now the special status should indicate that instead?

File: crates/bindings-csharp/Runtime/Internal/Module.cs [hunk 9]
```diff
@@ -233,29 +233,29 @@ private static byte[] Consume(this BytesSource source)
             // Write into the spare capacity of the buffer.
             var spare = buffer.AsSpan((int)written);
             var buf_len = (uint)spare.Length;
-            var ret = FFI._bytes_source_read(source, spare, ref buf_len);
+            var ret = (FFI.Errno)FFI._bytes_source_read(source, spare, ref buf_len);
```
Comment (RReverser) [inline crates/bindings-csharp/Runtime/Internal/Module.cs:236]: I'd change `bytes_source_read` to return `Errno` directly instead, to avoid casting.

File: crates/bindings-csharp/Runtime/Internal/ITable.cs (hunk 8 above)
Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/ITable.cs:48]: Right, but what I meant is what if `ret == 0 && buffer_len == 0`? Right now the iterator will stop on that too - is that something we want intentionally or do we want to stop _only_ on `ret == -1`?

Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/ITable.cs:48]: I was thinking of an edge case where host - for whatever reason - returned one of the interim buffers as genuinely empty, so `buffer_len == 0` but the iterator is not over yet.

If you think it should be unreachable, let's add `Debug.Assert` to indicate that and to check for that at least in debug builds.

But even with the assertion it still sounds like this particular line can be simplified to `return true;` as the "iterator finished" condition only depends on `handle == FFI.RowIter.INVALID` (after having received "exhausted" return code) and not depend on `buffer_len` at all.

File: crates/core/src/host/wasmtime/wasm_instance_env.rs (hunk 1 above)
Comment (RReverser) [reply crates/core/src/host/wasmtime/wasm_instance_env.rs:639]: Yeah either way its a future TODO, just wanted to comment with an alternative :)

File: crates/bindings-csharp/Runtime/Internal/ITable.cs (hunk 6 above)
Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/ITable.cs]: Ah yeah you can, in this case it was just unclear whether the fallthrough is intentional or not.

Nitpicking as the comment is enough but I'd probably just rewrite it to `case Errno.EXHAUSTED or Errno.OK:` like you said, which would make the intent clear instantly.

File: crates/bindings-csharp/Runtime/Internal/ITable.cs (hunk 8 above)
Comment (RReverser) [reply crates/bindings-csharp/Runtime/Internal/ITable.cs:51]: > Counter-example: Consider an empty table. You do the equivalent of `MyTable::iter()`. This makes the host side iterator contain zero chunks. A call to `_row_iter_bsatn_advance` will return `-1` the first time and you don't know that `handle == FFI.RowIter.INVALID` yet. Yet, the call to `.MoveNext()` should return `false`, as otherwise, the `foreach` will think there was a chunk/rows, but there wasn't (technically, `while (stream.Position < stream.Length)` is tolerant to this).

You're right, that's actually a great counter-example.

Will `Debug.Assert` suggested above fail on this? Either way, could be nice to have some test for this edge case. I'm guessing you'd like to do that outside this PR, but if so, can you at least create an issue so that we don't forget to add such SDK test later?

**Review by RReverser — APPROVED** [review]

File: crates/bindings-sys/src/lib.rs [hunk 10]
```diff
@@ -355,6 +380,11 @@ pub mod raw {
     #[repr(transparent)]
     pub struct RowIter(u32);
 
+    impl RowIter {
+        /// An invalid handle, used e.g., when the iterator has been exhausted.
+        pub const INVALID: Self = Self(0);
+    }
```
Comment (coolreader18) [inline crates/bindings-sys/src/lib.rs:386]: Would it make more sense to have RowIter be a `NonZeroU32`, and use `Option<RowIter>` to represent `INVALID`?
